File: BiliRecommendUI/src/main/java/com/bilirecommendui/main/widget/adapter/RecommendAdapter.java
package com.bilirecommendui.main.widget.adapter;


import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.widget.FrameLayout;

import com.bannerlayout.model.BannerModel;
import com.bilirecommendui.R;
import com.bilirecommendui.base.BaseRecyclerAdapter;
import com.bilirecommendui.base.SuperViewHolder;
import com.bilirecommendui.main.model.RecommendCompat;
import com.bilirecommendui.main.model.RecommendModel;
import com.bilirecommendui.main.widget.holder.RecommendFooterViewFactory;
import com.bilirecommendui.main.widget.holder.RecommendHeaderHolder;
import com.bilirecommendui.main.widget.holder.RecommendItemViewHolder;
import com.bilirecommendui.network.Constant;
import com.bilirecommendui.utils.ImageLoaderUtils;

import java.util.ArrayList;
import java.util.List;


/**
 * by y on 2016/9/17.
 */
public class RecommendAdapter extends BaseRecyclerAdapter<RecommendModel> {

    private List<BannerModel> banner = null;
    private List<RecommendCompat> compats = new ArrayList<>();

    @Override
    public int getItemViewType(int position) {
        return compats.get(position).getType();
    }

    @Override
    protected int getLayoutId(int viewType) {
        switch (viewType) {
            case RecommendCompat.TYPE_BANNER:
                return R.layout.item_recommend_banner;
            case RecommendCompat.TYPE_HEADER:
                return R.layout.item_recommend_header;
            case RecommendCompat.TYPE_FOOTER:
                return R.layout.item_recommend_footer;
            case RecommendCompat.TYPE_WEB_LINK:
                return R.layout.item_recommend_topic;
            case RecommendCompat.TYPE_ACTIVITY:
                return R.layout.item_recommend_activity;
            default:
                return R.layout.item_recommend_item;
        }
    }

    @Override
    public void onBindViewHolder(SuperViewHolder holder, int position) {
        if (null == banner || isNull()) {
            return;
        }
        int newPosition = compats.get(position).getNewPosition();
        switch (compats.get(position).getType()) {
            case RecommendCompat.TYPE_BANNER:
                holder
                        .getBannerLayout(R.id.banner)
                        .initListResources(banner)
                        .initTips().
                        start(true);
                break;
            case RecommendCompat.TYPE_HEADER:
                new RecommendHeaderHolder(holder).setHeaderData(mDatas.getResult(), newPosition);
                break;
            case RecommendCompat.TYPE_FOOTER:
                FrameLayout footView = holder.getFrameLayout(R.id.fl_recommend_footer);
                footView.removeAllViews();
                footView.addView(RecommendFooterViewFactory.createView(mDatas.getResult().get(newPosition)));
                break;
            case RecommendCompat.TYPE_ITEM:
                new RecommendItemViewHolder(holder).setData(mDatas.getResult().get(newPosition).getBody().get(compats.get(position).getItemPosition()), mDatas.getResult().get(newPosition).getType());
                break;
            case RecommendCompat.TYPE_ACTIVITY:
                RecyclerView recyclerView = holder.getRecyclerView(R.id.recyclerView);
                recyclerView.setHasFixedSize(true);
                recyclerView.setLayoutManager(new StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.HORIZONTAL));
                recyclerView.setAdapter(new RecommendActivityAdapter(mDatas.getResult().get(newPosition).getBody()));
                break;
            case RecommendCompat.TYPE_WEB_LINK:
                ImageLoaderUtils.display(holder.getImageView(R.id.iv_topic), mDatas.getResult().get(newPosition).getBody().get(0).getCover());
                break;
        }
    }


    @Override
    public int getItemCount() {
        int temp = 0;
        if (isNull()) {
            return temp;
        }
        compats.clear();
        compats.add(new RecommendCompat(temp += 1, RecommendCompat.TYPE_BANNER, -1));
        List<RecommendModel.ResultBean> result = mDatas.getResult();
        for (int i = 0; i < result.size(); i++) {
            switch (result.get(i).getType()) {
                case Constant.TYPE_RECOMMEND:
                case Constant.TYPE_LIVE:
                case Constant.TYPE_BANGUMI:
                case Constant.TYPE_REGION:
                    compats.add(new RecommendCompat(temp += 1, RecommendCompat.TYPE_HEADER, i));
                    for (int j = 0; j < result.get(i).getBody().size(); j++) {
                        compats.add(new RecommendCompat(temp += 1, RecommendCompat.TYPE_ITEM, i, j));
                    }
                    compats.add(new RecommendCompat(temp += 1, RecommendCompat.TYPE_FOOTER, i));
                    break;
                case Constant.TYPE_WEB_LINK:
                    compats.add(new RecommendCompat(temp += 1, RecommendCompat.TYPE_HEADER, i));
                    compats.add(new RecommendCompat(temp += 1, RecommendCompat.TYPE_WEB_LINK, i));
                    break;
                case Constant.TYPE_ACTIVITY_:
                    compats.add(new RecommendCompat(temp += 1, RecommendCompat.TYPE_HEADER, i));
                    compats.add(new RecommendCompat(temp += 1, RecommendCompat.TYPE_ACTIVITY, i));
                    break;
            }
        }
        return temp;
    }

    public void setBannerData(List<BannerModel> data) {
        this.banner = data;
    }

}

File: BilibiliRecommendUi/app/src/main/java/com/bilibilirecommend/main/widget/adapter/RecommendAdapter.kt
package com.bilibilirecommend.main.widget.adapter


import android.support.v7.widget.StaggeredGridLayoutManager
import com.bannerlayout.model.BannerModel
import com.bilibilirecommend.R
import com.bilibilirecommend.base.BaseRecyclerAdapter
import com.bilibilirecommend.base.SuperViewHolder
import com.bilibilirecommend.main.model.RecommendCompat
import com.bilibilirecommend.main.model.RecommendModel
import com.bilibilirecommend.main.widget.holder.RecommendFooterViewFactory
import com.bilibilirecommend.main.widget.holder.RecommendHeaderHolder
import com.bilibilirecommend.main.widget.holder.RecommendItemViewHolder
import com.bilibilirecommend.network.Constant
import com.bilibilirecommend.utils.ImageLoaderUtils
import java.util.*


class RecommendAdapter : BaseRecyclerAdapter() {

    private lateinit var mDatas: RecommendModel
    private lateinit var banner: List<BannerModel>
    private val compats = ArrayList<RecommendCompat>()

    override fun getItemViewType(position: Int): Int {
        return compats[position].type
    }

    fun refreshData(model: RecommendModel) {
        this.mDatas = model
        notifyDataSetChanged()
    }

    override fun getLayoutId(viewType: Int): Int {
        return when (viewType) {
            RecommendCompat.TYPE_BANNER -> R.layout.item_recommend_banner
            RecommendCompat.TYPE_HEADER -> R.layout.item_recommend_header
            RecommendCompat.TYPE_FOOTER -> R.layout.item_recommend_footer
            RecommendCompat.TYPE_WEB_LINK -> R.layout.item_recommend_topic
            RecommendCompat.TYPE_ACTIVITY -> R.layout.item_recommend_activity
            else -> R.layout.item_recommend_item
        }
    }

    override fun onBindViewHolder(holder: SuperViewHolder, position: Int) {
        if (!::banner.isInitialized) {
            return
        }
        val newPosition = compats[position].newPosition
        when (compats[position].type) {
            RecommendCompat.TYPE_BANNER -> holder
                    .getBannerLayout(R.id.banner)
                    .initListResources(banner)
                    .initTips().start(true)
            RecommendCompat.TYPE_HEADER -> RecommendHeaderHolder(holder).setHeaderData(mDatas.result, newPosition)
            RecommendCompat.TYPE_FOOTER -> {
                val footView = holder.getFrameLayout(R.id.fl_recommend_footer)
                footView.removeAllViews()
                footView.addView(RecommendFooterViewFactory.createView(mDatas.result[newPosition]))
            }
            RecommendCompat.TYPE_ITEM -> RecommendItemViewHolder(holder).setData(mDatas.result[newPosition].body[compats[position].itemPosition], mDatas.result[newPosition].type)
            RecommendCompat.TYPE_ACTIVITY -> {
                val recyclerView = holder.getRecyclerView(R.id.recyclerView)
                recyclerView.setHasFixedSize(true)
                recyclerView.layoutManager = StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.HORIZONTAL)
                recyclerView.adapter = RecommendActivityAdapter(mDatas.result[newPosition].body)
            }
            RecommendCompat.TYPE_WEB_LINK -> ImageLoaderUtils.display(holder.getImageView(R.id.iv_topic), mDatas.result[newPosition].body[0].cover)
        }
    }


    override fun getItemCount(): Int {
        var temp = 0
        if (!::mDatas.isInitialized) {
            return temp
        }
        compats.clear()
        temp += 1
        compats.add(RecommendCompat(temp, RecommendCompat.TYPE_BANNER, -1))
        val result = mDatas.result
        for (i in result.indices) {
            when (result[i].type) {
                Constant.TYPE_RECOMMEND, Constant.TYPE_LIVE, Constant.TYPE_BANGUMI, Constant.TYPE_REGION -> {
                    temp += 1
                    compats.add(RecommendCompat(temp, RecommendCompat.TYPE_HEADER, i))
                    for (j in 0 until result[i].body.size) {
                        temp += 1
                        compats.add(RecommendCompat(temp, RecommendCompat.TYPE_ITEM, i, j))
                    }
                    temp += 1
                    compats.add(RecommendCompat(temp, RecommendCompat.TYPE_FOOTER, i))
                }
                Constant.TYPE_WEB_LINK -> {
                    temp += 1
                    compats.add(RecommendCompat(temp, RecommendCompat.TYPE_HEADER, i))
                    temp += 1
                    compats.add(RecommendCompat(temp, RecommendCompat.TYPE_WEB_LINK, i))
                }
                Constant.TYPE_ACTIVITY_ -> {
                    temp += 1
                    compats.add(RecommendCompat(temp, RecommendCompat.TYPE_HEADER, i))
                    temp += 1
                    compats.add(RecommendCompat(temp, RecommendCompat.TYPE_ACTIVITY, i))
                }
            }
        }
        return temp
    }

    fun setBannerData(data: List<BannerModel>) {
        this.banner = data
    }
}
